### PR TITLE
hcm: prevent filter continuation after reset

### PR DESCRIPTION
**Pull request by snowp**
hcm: prevent filter continuation after reset

Adds a check to see if encoding has stopped before continuing with
encoding, ensuring that we don't continue encoding once the stream has
been reset. Adds tests for both the decoding and encoding path.

Risk Level: Medium
Testing: New integration tests
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

**Comment by mattklein123** [comment]
Check CI?

/wait

File: source/common/http/filter_manager.cc [hunk 1]
```diff
@@ -1288,6 +1288,8 @@ absl::optional<Router::ConfigConstSharedPtr> ActiveStreamDecoderFilter::routeCon
   return parent_.filter_manager_callbacks_.routeConfig();
 }
 
+bool ActiveStreamEncoderFilter::canContinue() { return !parent_.state_.encoding_complete_; }
```
Comment (alyssawilk) [inline source/common/http/filter_manager.cc:1291]: can we comment somewhere why we use local complete on the decoder path  and encoding complete on the encoder path?

File: source/common/http/filter_manager.h [hunk 2]
```diff
@@ -812,16 +815,17 @@ class FilterManager : public ScopeTrackedObject,
...
 
     uint32_t filter_call_state_{0};
 
     bool remote_complete_ : 1;
     bool local_complete_ : 1; // This indicates that local is complete prior to filter processing.
-                              // A filter can still stop the stream from being complete as seen
-                              // by the codec.
+    // A filter can still stop the stream from being complete as seen
+    // by the codec.
+    bool encoding_complete_ : 1;
```
Comment (alyssawilk) [inline source/common/http/filter_manager.h:828]: this comment made me think that encoding complete would go from true to false, if some filter retroactively made the stream not complete, where I *think* what actually happens is sometimes we set local_complete_ true before we set encoding_complete_

Maybe we can update the comments a bit to explain the difference more clearly?

**Review by alyssawilk — COMMENTED** [review]
code and test LGTM but I'd like to make the code a bit more clear before we land!

**Comment by snowp** [comment]
Seems like the TSAN test failure is real (can repro locally) but it's giving no output at all and I can't seem to get any information out of it (I'm guessing something is invoke std::abort or similar?). 

Will keep digging, but if anyone has any ideas I'm open to suggestions :)

**Pull request closed without merging**